File: androidsupport/src/main/java/com/jarklee/androidsupport/lifecycle/LifeCycleCollection.java
/*
 * ******************************************************************************
 *  Copyright Ⓒ 2017. TrinhQuan. All right reserved
 *  Author: TrinhQuan. Created on 2017/1/6
 *  Contact: trinhquan.171093@gmail.com
 * ******************************************************************************
 */

package com.jarklee.androidsupport.lifecycle;


import com.jarklee.androidsupport.callback.LifeCycleListener;

import java.util.LinkedList;
import java.util.List;

class LifeCycleCollection implements LifeCycleListener {

    private final List<LifeCycleListener> mCycleList;

    LifeCycleCollection() {
        mCycleList = new LinkedList<>();
    }

    int size() {
        return mCycleList.size();
    }

    void add(LifeCycleListener listener) {
        mCycleList.add(listener);
    }

    void remove(LifeCycleListener listener) {
        mCycleList.remove(listener);
    }

    private List<LifeCycleListener> getCycleList() {
        return mCycleList;
    }

    @Override
    public void onReady(LifeCycleHook hook) {
        //
    }

    @Override
    public void onCreate() {
        List<LifeCycleListener> cycleList = getCycleList();
        for (LifeCycleListener lifeCycleListener : cycleList) {
            lifeCycleListener.onCreate();
        }
    }

    @Override
    public void onStart() {
        List<LifeCycleListener> cycleList = getCycleList();
        for (LifeCycleListener lifeCycleListener : cycleList) {
            lifeCycleListener.onStart();
        }
    }

    @Override
    public void onResume() {
        List<LifeCycleListener> cycleList = getCycleList();
        for (LifeCycleListener lifeCycleListener : cycleList) {
            lifeCycleListener.onResume();
        }
    }

    @Override
    public void onPause() {
        List<LifeCycleListener> cycleList = getCycleList();
        for (LifeCycleListener lifeCycleListener : cycleList) {
            lifeCycleListener.onPause();
        }
    }

    @Override
    public void onStop() {
        List<LifeCycleListener> cycleList = getCycleList();
        for (LifeCycleListener lifeCycleListener : cycleList) {
            lifeCycleListener.onStop();
        }
    }

    @Override
    public void onDestroy() {
        List<LifeCycleListener> cycleList = getCycleList();
        for (LifeCycleListener lifeCycleListener : cycleList) {
            lifeCycleListener.onDestroy();
        }
    }
}

File: androidsupport-kotlin/src/main/java/com/jarklee/androidsupport/lifecycle/LifeCycleCollection.kt
/*
 * ******************************************************************************
 *  Copyright Ⓒ 2016. TrinhQuan. All right reserved
 *  Author: TrinhQuan. Created on 2016/12/20
 *  Contact: trinhquan.171093@gmail.com
 * ******************************************************************************
 */

package com.jarklee.androidsupport.lifecycle

import com.jarklee.androidsupport.callback.LifeCycleListener
import com.jarklee.androidsupport.lifecycle.LifeCycleHook

import java.util.LinkedList

internal class LifeCycleCollection : LifeCycleListener {

    private val mCycleList: MutableList<LifeCycleListener>

    init {
        mCycleList = LinkedList<LifeCycleListener>()
    }

    fun size(): Int {
        return mCycleList.size
    }

    fun add(listener: LifeCycleListener) {
        mCycleList.add(listener)
    }

    fun remove(listener: LifeCycleListener) {
        mCycleList.remove(listener)
    }

    private val cycleList: List<LifeCycleListener>
        get() = mCycleList

    override fun onReady(hook: LifeCycleHook) {
        //
    }

    override fun onCreate() {
        val cycleList = cycleList
        for (lifeCycleListener in cycleList) {
            lifeCycleListener.onCreate()
        }
    }

    override fun onStart() {
        val cycleList = cycleList
        for (lifeCycleListener in cycleList) {
            lifeCycleListener.onStart()
        }
    }

    override fun onResume() {
        val cycleList = cycleList
        for (lifeCycleListener in cycleList) {
            lifeCycleListener.onResume()
        }
    }

    override fun onPause() {
        val cycleList = cycleList
        for (lifeCycleListener in cycleList) {
            lifeCycleListener.onPause()
        }
    }

    override fun onStop() {
        val cycleList = cycleList
        for (lifeCycleListener in cycleList) {
            lifeCycleListener.onStop()
        }
    }

    override fun onDestroy() {
        val cycleList = cycleList
        for (lifeCycleListener in cycleList) {
            lifeCycleListener.onDestroy()
        }
    }
}
